agent.py

```diff
@@ -156,13 +157,133 @@ def __get_action(self, new_state):
             return np.random.randint(self.__action_space_len - 1)
         else:
             return self.__q.policy(new_state)
+
+    @property
+    def info(self) -> dict:
+        return {
+            **super().info,
+            "discretizer": self.__discretizer.info,
+        }
+
+
+class DQNAgent(Agent):
+    TYPE = AgentType.DQN
+    
+    def __init__(self, ctx: AgentContext, nn: DQN):
+        super().__init__(ctx)
+        self.nn = nn
+        self.steps = 0
+        self.grid_size = ctx.size
+        self.alpha = ctx.alpha
+        self.gamma = ctx.gamma
+
+        self.channels = 3
+        self.memory_size = 100_000
+        self.batch_size = 256
+        self.T = 1100
+        self.epsilon_start = 0.9
+        self.epsilon_end = 0.1
+        self.epsilon_decay = 500
+
+        self.policy_net = self.nn
+        self.target_net = deepcopy(self.nn)
+
+        self.replay_memory = ReplayMemory(self.memory_size, self.channels, self.grid_size)
+        self.tensorized_obs = self.__tensorize_observation(ctx.env.reset())
+
+        self.policy_net.train(True)
+        self.target_net.train(False)
         
-    def run_episode(self):
-        return super().run_episode()
+        self.policy_net.init_layers()
+        self.__copy_q_to_target()
+
+        self.optimizer = torch.optim.AdamW(self.policy_net.parameters(), lr=self.alpha)
+
+    def get_optimal_action(self, observation) -> int:
+        observation = self.__tensorize_observation(observation)
+        with torch.no_grad():
+            action_q_vals = self.policy_net(observation.view(1, self.channels, self.grid_size, self.grid_size)) 
+        return torch.argmax(action_q_vals).item()
+
+    def initialize(self):
+        self.__experience_initial()
```

Comment:
This wrapper is unnecessary? The "__experience_initial" function's functionality should be moved here

Reply:
Fixed, good spot!